torch/package/importer.py

```diff
@@ -448,3 +459,48 @@ def patched_getfile(object):
             return _package_imported_modules[object.__module__].__file__
     return _orig_getfile(object)
 inspect.getfile = patched_getfile
+
+
+class _PackageResourceReader:
+    """Private class used to support PackageImporter.get_resource_reader().
+
+    Confirms to the importlib.abc.ResourceReader interface. Allowed to access
+    the innards of PackageImporter.
+    """
+    def __init__(self, importer, fullname):
+        self.importer = importer
+        self.fullname = fullname
+
+    def open_resource(self, resource):
+        from io import BytesIO
+        return BytesIO(self.importer.load_binary(self.fullname, resource))
+
+    def resource_path(self, resource):
+        # The contract for resource_path is that it either returns a concrete
+        # file system path or raises FileNotFoundError.
+        raise FileNotFoundError
+
+    def is_resource(self, name):
+        path = self.importer._zipfile_path(self.fullname, name)
+        return self.importer.zip_reader.has_record(path)
+
+    def contents(self):
```

Comment:
If this is written another way it can avoid having to add boolean flags to `_zipfile_path`, which I think it is an improvement:

```
filename = self.fullname.replace('.', '/')
for f in get_all_records()
   match = re.match(f"^{quote(self.filename)}/([^/]+)$", f)
  if match:
    yield match[0]
```

Reply:
tried this, but in order get conformance with resource_reader API the regex gets complicated enough to not be immediately clear (in order to capture subdirs, future `/` slashes are optionally allowed but are not part of the match group). Seems more readable for future people to leave it as is imo, plus this is the implementation I stole from cpython lol